src/observables/observable_udp_socket.c

```diff
@@ -30,29 +30,24 @@ static void on_receive(uv_udp_t *handle, ssize_t nread, const uv_buf_t *buf, con
                        unsigned flags) {
     Observable *observable = (Observable *)handle->data;
 
-    if (addr == NULL) {
-        // FIXME: do not know why libuv does this. addr == NULL, nread == 0.
-        // We do not need to react on this.
+    if (nread == 0 && addr != NULL) { // an empty frame received
+        // do nothing, just for documentation
+    } else if (nread == 0 && addr == NULL) { // end of data
         observable->callback(observable, end_of_data());
-        return;
-    }
-
-    if (nread < 0) {
+    } else if (nread < 0) { // some error occured
         log_error("Failed to receive data from socket because of %s.", uv_strerror(nread));
-
-        free(buf->base);
-        return;
+    } else { // normal data received
+        // BEWARE: put a (char *) to user. May be dangerous,
+        // because it should not be a null-terminated C string.
+        void *data = observable->callback(observable, (void *)buf->base);
+
+        if (data) {
+            observable_broadcast(observable, data);
+        }
     }
 
-    // BEWARE: put a (char *) to user. May be dangerous,
-    // because it should not be a null-terminated C string.
-    void *data = observable->callback(observable, (void *)buf->base);
-
-    if (data) {
-        observable_broadcast(observable, data);
-    }
-
-    free(buf->base);
+    if (buf->base)
```

Comment:
Same with curly braces. We always should try to use condition body to avoid heisenbugs.